src/solver/coeffs.jl

```diff
@@ -0,0 +1,324 @@
+    oper = operation(expr)
+    expr = term(oper, args...)
+    return subs, expr
+end
+
+
+"""
+    filter_poly(og_expr, var)
+
+Filters polynomial expressions from weird irrationals that mess up
+the functionality of other helper functions for our solvers such as
+polynomial_coeffs and factor_use_nemo.
+
+# Arguments
+- og_expr: Original passed expression. This is deepcopied so that the
+user's expression is not altered unwillingly.
+- var: Var that the expression should be filtered with respect to.
+
+# Examples
+```jldoctest
+julia> filter_poly(x + 2im, x)
+(Dict{Any, Any}(var"##244" => im), 2var"##244" + x)
+
+julia> filter_poly((1/im)*x + 3*y*z, x)
+(Dict{Any, Any}(var"##245" => -1.0, var"##246" => im), 3y*z + var"##245"*var"##246"*x)
+
+julia> filter_poly((x+1)*term(log, 3), x)
+(Dict{Any, Any}(var"##247" => log(3)), var"##247"*(1 + x))
+```
+"""
+function filter_poly(og_expr, var)
+    expr = deepcopy(og_expr)
+    expr = unwrap(expr)
+    vars = get_variables(expr)
+
+    # handle edge cases
+    if !isequal(vars, []) && isequal(vars[1], expr)
+        return (Dict{Any, Any}(), expr)
+    elseif isequal(vars, [])
+        return filter_stuff(expr)
+    end
+
+    # core filter
+    subs, expr = _filter_poly(expr, var)
+
+    # reassemble expr to avoid variables remembering original values issue and clean
+    args = arguments(expr)
+    oper = operation(expr)
+    new_expr = clean_f(term(oper, args...), var, subs)
+
+    return subs, new_expr
+end
+
+
+"""
+    sdegree(coeffs, var)
+Gets the degree of a polynomial by traversing the `coeffs`
+output from polynomial_coeffs.
+
+# Arguments
+- coeffs: output from polynomial_coeffs
+- var: var present in polynomial_coeffs
+
+# Examples
+```jldoctest
+julia> coeffs, constant = polynomial_coeffs(x^2 + x + 1, [x])
+(Dict{Any, Any}(x^2 => 1, x => 1, 1 => 1), 0)
+
+julia> sdegree(coeffs, x)
+2
+
+
+julia> coeffs, constant = polynomial_coeffs(x^12 + 3x^2, [x])
+(Dict{Any, Any}(x^2 => 3, x^12 => 1), 0)
+
+julia> sdegree(coeffs, x)
+12
+```
+"""
+function sdegree(coeffs, var)
+    degree = 0
+    vars = collect(keys(coeffs))
+    for n in vars
+        isequal(n, 1) && continue
+        isequal(n, var) && degree > 1 && continue
+
+        if isequal(n, var) && degree < 1 
+            degree = 1
+            continue
+        end
+
+        args = arguments(n)
+        if args[2] > degree 
+            degree = args[2]
+        end
+    end
+    return degree
+end
+
+
+
+########## NOT USED ########## 
+
+function lead_term(expr, var)
+    subs, expr = filter_poly(expr, var)
+    coeffs, constant = polynomial_coeffs(expr, [var])
+    degree = degree(expr, var)
+    lead_term = coeffs[var^degree]*var^degree
+    for (var, sub) in subs
+        lead_term = substitute(lead_term, Dict([var => sub]), fold=false)
+    end
+
+    return lead_term
+end
+
+function lead_coeff(expr, var)
+    degree = degree(expr, var)
+    lead_coeff = lead_term(expr, var) / (var^degree)
+    return lead_coeff
+end
+
```

Comment:
To be removed then?

Reply:
May be used for the future when implementing parametric  multivar solver. Will mark as resolved for now.